File: app/src/main/java/com/sdin/tourstamprally/ui/fragment/LocationFragment.java
package com.sdin.tourstamprally.ui.fragment;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.location.Address;
import android.location.Geocoder;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.databinding.DataBindingUtil;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.LinearLayout;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.sdin.tourstamprally.R;
import com.sdin.tourstamprally.Utils;
import com.sdin.tourstamprally.databinding.FragmentLocationBinding;
import com.sdin.tourstamprally.databinding.ItemReRallyMapBinding;
import com.sdin.tourstamprally.databinding.LocationReItemBinding;
import com.sdin.tourstamprally.model.Location_four;
import com.sdin.tourstamprally.model.RallyMapDTO;
import com.sdin.tourstamprally.model.Tour_Spot;
import com.sdin.tourstamprally.model.TouristSpotPoint;
import com.sdin.tourstamprally.ui.activity.MainActivity;
import com.sdin.tourstamprally.ui.dialog.GuidDialog;
import com.sdin.tourstamprally.utill.ItemOnClick;
import com.sdin.tourstamprally.utill.ItemOnClickAb;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import gun0912.tedkeyboardobserver.TedKeyboardObserver;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LocationFragment extends BaseFragment {

    private FragmentLocationBinding binding;
    private String location_name;
    private Location_four location_four;
    private List<RallyMapDTO> list;
    private Map<Integer, Integer> spot_poinMap;
    private Map<Integer, Integer> spot_HistoryMap;


    public static LocationFragment newInstance(Location_four location_four) {

        Bundle args = new Bundle();

        LocationFragment fragment = new LocationFragment();
        //args.putSerializable("model", location_four);
        args.putParcelable("model", location_four);
        fragment.setArguments(args);
        Log.wtf("Locationa newInstance", String.valueOf(MainActivity.keyboardState));
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            //location_four = (Location_four) getArguments().getSerializable("model");
            location_four = getArguments().getParcelable("model");
            location_name = location_four.getLocation_name();
            Log.wtf("Locationa", String.valueOf(MainActivity.keyboardState));


        }
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_location, container, false);
        binding.locationTxv.setText(location_name);


        if (!TextUtils.isEmpty(location_four.getLocation_img())) {
            Glide.with(requireContext()).load("http://coratest.kr/imagefile/bsr/" + location_four.getLocation_img()).into(new CustomTarget<Drawable>() {
                @Override
                public void onResourceReady(@NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
                    binding.topLayout.setBackground(resource);

                }

                @Override
                public void onLoadCleared(@Nullable Drawable placeholder) {

                }
            });
        }

        getData();
        return binding.getRoot();
    }

    private void getData() {
        binding.locationPgb.setVisibility(View.VISIBLE);
        apiService.getTourLocation_for_spot(Utils.User_Idx, location_four.getLocation_idx()).enqueue(new Callback<List<RallyMapDTO>>() {
            @Override
            public void onResponse(@NotNull Call<List<RallyMapDTO>> call, @NotNull Response<List<RallyMapDTO>> response) {
                list = response.body();
                setRecylcerviewAdapter();
            }

            @Override
            public void onFailure(@NotNull Call<List<RallyMapDTO>> call, @NotNull Throwable t) {

            }
        });


    }

    private void setRecylcerviewAdapter() {

        LocationFragAdapter adapter = new LocationFragAdapter(new ArrayList<>(list));

        binding.recyclerviewLocationRe.setAdapter(adapter);
        binding.recyclerviewLocationRe.setLayoutManager(new GridLayoutManager(requireContext(), 2));
        setProgress();
    }

    @SuppressLint("SetTextI18n")
    private void setProgress() {

        //int clear = 0;

        int AllLocationPercent = 0;
        int MyLocationPercent = 0;
        for (RallyMapDTO model : list) {
            AllLocationPercent += model.getAllCount();
            MyLocationPercent += model.getMyCount();
        }
        int allCountd = (int) ((double) MyLocationPercent / AllLocationPercent * 100);

        binding.seekBarLocation.setMax(AllLocationPercent);
        binding.seekBarLocation.setProgress(MyLocationPercent);
        binding.seekPercentTxv.setText(allCountd + "%");
        binding.locationPgb.setVisibility(View.GONE);

    }

    class LocationFragAdapter extends RecyclerView.Adapter<LocationFragAdapter.ViewHolder> {

        private final ArrayList<RallyMapDTO> arrayList;
        //private final ItemOnClick listener;
        private RallyMapDTO send_model;
        private Geocoder geocoder;


        public LocationFragAdapter(ArrayList<RallyMapDTO> arrayList) {
            this.arrayList = arrayList;
            //listener = (MainActivity) requireActivity();
            geocoder = new Geocoder(requireContext());
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

            return new ViewHolder(
                    ItemReRallyMapBinding.inflate(
                            LayoutInflater.from(requireContext()), parent, false)
            );
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

            Glide.with(holder.itemView.getContext()).load(position % 2 == 0 ? R.drawable.icon_deep_blue : R.drawable.icon_sky_blue).into(holder.binding.locationImv);
            Glide.with(holder.itemView.getContext()).load(setProgress(position) ? R.drawable.stemp_ic : R.drawable.logo_gray).into(holder.binding.logoImv);
            holder.binding.spotName.setText(arrayList.get(position).getTouristspot_name());
            if (arrayList.get(position).getTouristspot_address() != null
                    && arrayList.get(position).getTouristspot_address().equalsIgnoreCase("null")) {
                holder.binding.explanTxv.setText(arrayList.get(position).getTouristspot_address());
            }

            Glide.with(holder.binding.spotImv.getContext())
                    .load("http://coratest.kr/imagefile/bsr/" + arrayList.get(position).getTouristspot_img())
                    /*.apply(RequestOptions.bitmapTransform(new RoundedCorners(25)))*/
                    .error(R.drawable.sample_bg)
                    .into(holder.binding.spotImv);

        }

        private boolean setProgress(int position) {

            return arrayList.get(position).getAllCount() <= arrayList.get(position).getMyCount();

        }

        @Override
        public int getItemCount() {
            return arrayList.size();
        }

        class ViewHolder extends RecyclerView.ViewHolder {
            private final ItemReRallyMapBinding binding;

            public ViewHolder(@NonNull ItemReRallyMapBinding binding) {
                super(binding.getRoot());
                this.binding = binding;

                binding.topLayout.setOnClickListener(v -> {
                    send_model = arrayList.get(getAbsoluteAdapterPosition());
                    Bundle bundle = new Bundle();
                    bundle.putSerializable("model", send_model);
                    bundle.putString("title", send_model.getTouristspot_name());
                    Navigation
                            .findNavController(requireActivity(), R.id.nav_host)
                            .navigate(R.id.action_fragment_location_to_fragment_tour_spot_point, bundle);
                    //listener.ItemGuidForPoint(send_model);
                    /*GuidDialog guidDialog = new GuidDialog(requireContext());
                    guidDialog.show();
                    guidDialog.setClickListener(itemOnClick);*/
                });
            }
        }


        /*private final ItemOnClick itemOnClick = new ItemOnClickAb() {
            @Override
            public void ItemGuid(int position) {
                Log.d("dialog Onclick Listener", String.valueOf(position));
                if (position == 1 || position == 2){
                    listener.ItemGuid(position);
                }else {
                    listener.ItemGuidForPoint(send_model);
                }
            }
        };*/
    }


}

File: app/src/main/java/com/sdin/tourstamprally/ui/fragment/tour/LocationFragment.kt
package com.sdin.tourstamprally.ui.fragment.tour

import android.annotation.SuppressLint
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.sdin.tourstamprally.R
import com.sdin.tourstamprally.Utils
import com.sdin.tourstamprally.databinding.FragmentLocationBinding
import com.sdin.tourstamprally.databinding.ItemReRallyMapBinding
import com.sdin.tourstamprally.model.RallyMapModel
import com.sdin.tourstamprally.model.TopFourLocationModel
import com.sdin.tourstamprally.ui.fragment.BaseFragment
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import kotlin.math.roundToInt

class LocationFragment : BaseFragment() {
    private var binding: FragmentLocationBinding? = null
    private lateinit var model: TopFourLocationModel
    private lateinit var list: List<RallyMapModel>
    private lateinit var locationFragmentAdapter: LocationFragAdapter

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_location, container, false)
        val args: LocationFragmentArgs by navArgs()
        model = args.topFourModel

        binding?.btnGetstamp?.setOnClickListener {
            val action = LocationFragmentDirections.actionFragmentLocationToPageStamp()
            findNavController().navigate(action)
        }
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstancdState: Bundle?) {
        binding?.apply {
            locationTxv.text = model.location_name

            Glide.with(topLayout.context)
                .load("http://coratest.kr/imagefile/bsr/" + model.location_img)
                .into(object : CustomTarget<Drawable>() {
                    override fun onResourceReady(
                        resource: Drawable,
                        transition: Transition<in Drawable>?
                    ) {
                        topLayout.background = resource
                    }

                    override fun onLoadCleared(placeholder: Drawable?) {

                    }

                })
        }

    }

    override fun onResume() {
        super.onResume()
        getData()
    }

    private fun getData() {
        binding!!.locationPgb.visibility = View.VISIBLE
        apiService.getTourLocation_for_spot(Utils.User_Idx, model.location_idx)
            .enqueue(object : Callback<List<RallyMapModel>> {
                override fun onResponse(
                    call: Call<List<RallyMapModel>>,
                    response: Response<List<RallyMapModel>>
                ) {
                    val result = response.body()
                    if (result != null) {
                        list = result
                    }

                    setRecyclerViewAdapter()
                }

                override fun onFailure(call: Call<List<RallyMapModel>>, t: Throwable) {
                    t.printStackTrace()
                    binding!!.locationPgb.visibility = View.GONE
                }
            })
    }

    private fun setRecyclerViewAdapter() {
        if (context == null) {
            binding?.locationPgb?.visibility = View.GONE
            return
        }

        binding?.recyclerviewLocationRe?.apply {
            locationFragmentAdapter = LocationFragAdapter(callback = { selectedModel ->
                val action =
                    LocationFragmentDirections.actionFragmentLocationToFragmentTourSpotPoint(
                        selectedModel.touristspot_name,
                        selectedModel
                    )


                findNavController().navigate(action)

            }, interCallback = {

                it.user_touristspot_interest_idx?.let { interIdx ->

                    if (interIdx > 0) {
                        deleteTourInter(it)
                    } else {

                        interTour(it)
                    }

                }


            }).apply {
                submitList(list)
            }

            adapter = locationFragmentAdapter

            layoutManager = GridLayoutManager(requireContext(), 2)
            binding!!.locationPgb.visibility = View.GONE
            setProgress()
        }
    }

    private fun interTour(model: RallyMapModel) {
        apiService.insert_intest(Utils.User_Idx, model.touristspot_idx)
            .enqueue(object : Callback<Int> {
                override fun onResponse(call: Call<Int>, response: Response<Int>) {
                    if (response.isSuccessful) {
                        val result = response.body()
                        if (result != null) {
                            if (result > 0) {

                                val currentList: MutableList<RallyMapModel> = mutableListOf()
                                currentList.addAll(locationFragmentAdapter.currentList)
                                val position = currentList.indexOf(model)
                                val subModel = model
                                subModel.user_touristspot_interest_idx = result
                                currentList[position] = subModel

                            }
                        }
                    }
                }

                override fun onFailure(call: Call<Int>, t: Throwable) {
                    t.printStackTrace()
                }

            })
    }

    private fun deleteTourInter(model: RallyMapModel) {
        model.user_touristspot_interest_idx?.let {
            apiService.remove_intest(it, 1).enqueue(object : Callback<Int> {
                override fun onResponse(call: Call<Int>, response: Response<Int>) {
                    if (response.isSuccessful) {
                        val result = response.body()
                        if (result != null) {
                            if (result > 0) {
                                val currentList: MutableList<RallyMapModel> = mutableListOf()
                                currentList.addAll(locationFragmentAdapter.currentList)
                                val position = currentList.indexOf(model)
                                val subModel = model
                                subModel.user_touristspot_interest_idx = 0
                                currentList[position] = subModel

                            }

                        }
                    }
                }

                override fun onFailure(call: Call<Int>, t: Throwable) {
                    t.printStackTrace()
                }

            })
        }
    }

    @SuppressLint("SetTextI18n")
    private fun setProgress() = with(binding!!) {
        if (model.myHistoryCount != 0 && model.allPointCount != 0) {
            val myCount =
                ((model.myHistoryCount.toDouble()) / model.allPointCount.toDouble() * 100).roundToInt()

            seekBarLocation.max = 100
            seekBarLocation.progress = myCount
            seekPercentTxv.text = "${myCount}%"
            locationPgb.visibility = View.GONE

        }
    }


    class LocationFragAdapter(
        private val callback: (RallyMapModel) -> Unit,
        private val interCallback: (RallyMapModel) -> Unit
    ) :
        ListAdapter<RallyMapModel, LocationFragAdapter.ViewHolder>(differ) {
        inner class ViewHolder(private val viewHolderBinding: ItemReRallyMapBinding) :
            RecyclerView.ViewHolder(viewHolderBinding.root) {

            fun onBind(model: RallyMapModel) {
                with(viewHolderBinding) {
                    Glide.with(locationImv.context)
                        .load(
                            if (absoluteAdapterPosition % 2 == 0)
                                R.drawable.location_puple
                            else R.drawable.location_sky
                        ).into(locationImv)

                    spotContainer.let {
                        if(absoluteAdapterPosition % 4 == 0 ||absoluteAdapterPosition % 4 == 1){
                            it.setBackgroundResource(R.drawable.bg_rounded_03_location)
                        }
                        else{
                            it.setBackgroundResource(R.drawable.bg_rounded_03)
                        }
                    }


                    Glide.with(logoImv.context)
                        .load(
                            if (setProgress(model = model))
                                R.drawable.stemp_ic
                            else R.drawable.logo_gray
                        ).into(logoImv)

                    spotName.text = model.touristspot_name

                    if (!model.touristspot_address.isNullOrEmpty()) {
                        explanTxv.text = model.touristspot_address
                    }

                    Glide.with(spotImv.context)
                        .load("http://coratest.kr/imagefile/bsr/" + model.touristspot_img)
                        .error(R.drawable.sample_bg)
                        .into(spotImv)

                    topLayout.setOnClickListener {
                        callback(model)
                    }

                    model.user_touristspot_interest_idx?.let {
                        val deepImg =
                            if (it <= 0) {
                                R.drawable.heart_resize
                            } else {
                                R.drawable.full_heart_resize
                            }

                        Glide.with(deepImb.context).load(deepImg).into(deepImb)
                    } ?: kotlin.run {
                        Glide.with(deepImb.context).load(R.drawable.heart_resize).into(deepImb)
                    }


                    deepImb.setOnClickListener {
                        model.user_touristspot_interest_idx?.let {
                            val heartImage =
                                if (it > 0) {
                                    R.drawable.heart_resize


                                } else {
                                    R.drawable.full_heart_resize
                                }
                            Glide.with(deepImb.context)
                                .load(heartImage)
                                .into(deepImb)

                            interCallback(model)

                        } ?: kotlin.run {
                            Glide.with(deepImb.context)
                                .load(R.drawable.full_heart_resize)
                                .into(deepImb)

                            interCallback(model)
                        }
                    }


                }
            }

            private fun setProgress(model: RallyMapModel): Boolean {
                if (model.allCount != null && model.myCount != null) {
                    return model.allCount <= model.myCount
                }
                return false

            }

        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(
                ItemReRallyMapBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
                )
            )
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.onBind(currentList[position])

        }
    }

    companion object {
        val differ = object : DiffUtil.ItemCallback<RallyMapModel>() {
            override fun areItemsTheSame(oldItem: RallyMapModel, newItem: RallyMapModel): Boolean {
                return oldItem.touristspot_idx == newItem.touristspot_idx
            }

            override fun areContentsTheSame(
                oldItem: RallyMapModel,
                newItem: RallyMapModel
            ): Boolean {
                return oldItem == newItem
            }

        }
    }
}